library/src/main/java/eightbitlab/com/blurview/BlurViewCanvas.java

```diff
@@ -2,13 +2,47 @@
 
 import android.graphics.Bitmap;
 import android.graphics.Canvas;
+import android.util.Log;
 
 import androidx.annotation.NonNull;
 
-// Servers purely as a marker of a Canvas used in BlurView
-// to skip drawing itself and other BlurViews on the View hierarchy snapshot
+/** Servers as a marker of a Canvas used in BlurView to skip drawing
+ * itself and other BlurViews on the View hierarchy snapshot
+ *
+ * Also used to avoid BlurView crashing when the {@code rootView} call
+ * {@link Canvas#save()} and {@link Canvas#restore()} an unequal amount of times.
+ * */
 public class BlurViewCanvas extends Canvas {
+    private int saveCountInitial;
+
     public BlurViewCanvas(@NonNull Bitmap bitmap) {
         super(bitmap);
     }
+
+    public void saveRoot() {
```

Comment:
I don't think the custom method is a good idea.
Some other code might still call regular `save` on this canvas, invalidating the whole fix.

Reply:
`saveRoot` is only used to get the save index, in fact there is no need to change the `save` function behavior.

If there is more `save` than `restore` the `restoreRoot` function will restore to the correct index.
If there is more `restore` than `save`, we need to keep the original transform intact, so this is why `restore` need to be overridden.

In fact, code calling the regular `save` is intended behavior of this PR as there is no need to change how it works.
In fact the `saveRoot` and `restoreRoot` should only be called by `BlurView`.

I will update the javadoc, sorry it is my fault for not making the code more understandable.

Reply:
Thank for reviewing the patch, I added the javadoc to make it easier to understand.

The initial version of the patch hooked `save` and `restore`, but it doesn't work because of `restoreToCount`.

If you have any more questions please feel free to ask.

Reply:
Alright, could you please explain how [this](https://github.com/Dimezis/BlurView/pull/187) happens exactly?
Because I'm struggling to understand how to reproduce this crash

Reply:
It's just bad code calling `save()` or `restore()` too much times, this commit goal is to make sure the crash does not come from `BlurView` when it is not `BlurView` fault.

This patch doesn't prevent crashes, it just make the stacktrace show the actual error instead of just showing it comes from `BlurView` when it has nothing to do with `BlurView`.

To reproduce this crash, it just require to create a `View` that call `restore()` in it's draw function without calling `save()`  
(Aka. making bad code)